Fix sorting and pagination in the `mc/price-benchmarks` endpoint

mukeshpanchal27 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #2911.

_Replace this with a good description of your changes & reasoning._


### Screenshots:

<!--- Optional --->


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

- Remove local DB `wp_gla_merchant_price_benchmarks` if you have already in DB.
- Add `add_filter( 'woocommerce_gla_force_run_install', '__return_true' );` filter in your mu-plugin to create DB table.
- Visit any admin page
- Check local DB `wp_gla_merchant_price_benchmarks` table is created or not.
- Testing REST API callback permission.
  - If using a service like PostMan, follow [these instructions](https://github.com/woocommerce/google-listings-and-ads/wiki/Testing-GLA-endpoints-with-Postman) to set up auth cookies and nonce.
  - To force the permissions to return true, you can use the following snippet:
  
```php
/**
 * Filters user capabilities to ensure `manage_woocommerce` capability always returns true.
 *
 * This function modifies the `$allcaps` array to grant the `manage_woocommerce` capability
 * to the current user if it is being checked.
 *
 * @param array $allcaps An array of all the capabilities of the user.
 * @param array $caps An array of capabilities being checked.
 * @return array Filtered array of all capabilities.
 */
function filter_allow_manage_woocommerce_caps( $allcaps, $caps ) {
	if ( isset( $caps[0] ) && 'manage_woocommerce' === $caps[0] ) {
		$allcaps[ $caps[0] ] = true;
	}
	return $allcaps;
}
add_filter( 'user_has_cap', 'filter_allow_manage_woocommerce_caps', 10, 2 );
```

- Use instruction from #2890 to set up the test-proxy to mock API responses
- Send a GET request to the /wc/gla/mc/price-benchmarks/ endpoint
- Review the response returned from the price-benchmarks endpoint
- Test schedule job.
  - Visit `Tools > Scheduled Actions` from the admin menu `/wp-admin/tools.php?page=action-scheduler` and verify that the `gla/jobs/update_merchant_price_benchmarks/process_item` job has been scheduled.
  - If the job hasn't already run, you can run it from the job list
  - Confirm that the data is imported to our local database



### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

>

mukeshpanchal27 committed:
Update DB schema

mukeshpanchal27 committed:
Separating the responsibility in get_price_benchmark_counts

mukeshpanchal27 committed:
Update changes for update_price_benchmarks

mukeshpanchal27 committed:
Update performance query

mukeshpanchal27 committed:
Introduce get_price_benchmarks_data()

mukeshpanchal27 committed:
Update effectiveness value

mukeshpanchal27 committed:
Update the method names to be more consistent

mukeshpanchal27 committed:
Update method name in unit tests

mukeshpanchal27 committed:
Update PriceBenchmarksController

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -21,6 +22,196 @@
+	 * Retrieves the product thumbnail URL.
+	 *
+	 * @param int $product_id WooCommerce product ID.
+	 * @return string|null Product thumbnail URL or null if not found.
```
The docs don't match the implementation

```suggestion
	 * @return string Product thumbnail URL or an empty string if not found.
```

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -21,6 +22,196 @@
+	 *                              namespaced product ID (`online:en:GB:merchant_center_id`)
+	 * @return int the ID for the WC product linked to the provided Google product ID (0 if not found)
+	 */
+	public function get_product_id( string $mc_product_id ): int {
```
Can we now remove this helper and replace it with the `ProductHelper::get_wc_product_id()` method?

mukeshpanchal27 replied:
Updated in https://github.com/woocommerce/google-listings-and-ads/pull/2935/commits/2749a502eec8bcde1e1d5c732a1552210ad35a6c

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -43,18 +234,36 @@ public function update_price_benchmarks(): void {
 			// Insert new benchmark data.
-			foreach ( $benchmarks as $benchmark ) {
+			foreach ( $benchmarks as $id => $benchmark ) {
+				$performance_data = $this->get_performance_data( $id );
```
Now that we're getting all the Metrics data at once, let's avoid individual API calls based on IDs and instead query all the `MerchantPerformanceView` data in a single query and map it to the `PriceCompetitivenessProductView` and `PriceInsightsProductView` in one process.

mukeshpanchal27 replied:
Addressed in https://github.com/woocommerce/google-listings-and-ads/pull/2935/commits/d26d2d0f09451c4043f13b751a62266b91fbea4e

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -43,18 +234,36 @@ public function update_price_benchmarks(): void {
-						'benchmark_price_micros'        => $benchmark['benchmark_price_micros'],
-						'benchmark_price_currency_code' => $benchmark['benchmark_price_currency_code'],
-						'price_compared_with_benchmark' => $price_compared_with_benchmark,
+						'product_id'                      => $benchmark['id'],
+						'mc_product_id'                   => $benchmark['id'],
```
`product_id` should be the WC product ID. `mc_product_id` should be the ID returned from the `PriceCompetitivenessProductView` or `PriceInsightsProductView` queries (e.g., `online:en:US:12345`).

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -43,18 +234,36 @@ public function update_price_benchmarks(): void {
+
 				$price_compared_with_benchmark = $this->price_compared_with_benchmark( $benchmark['price_micros'], $benchmark['benchmark_price_micros'] );
+				$effectiveness                 = $this->get_effectiveness( $benchmark['effectiveness'] );
 				$query->insert(
```
This whole insert could be simplified a lot if we match and normalize the column keys during the mapping process, that way we could simply do something like: 

```php
foreach( $benchmarks as $benchmark_item ) {
  $query->insert( $benchmark_item );
}
```

mukeshpanchal27 replied:
Nice catch. it look more simplified.

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -101,15 +310,145 @@ public function get_summary(): array {
+
+		if ( ! empty( $args['search'] ) ) {
+			$search = trim( $args['search'] );
+			// TODO: Implement a search mechanism.
```
We could consider adding product title to the price benchmarks DB if it's easier, or search the products to get a list of IDs and then use those IDs like `$query->where( 'product_id', $results, 'IN' )`

mukeshpanchal27 replied:
If we store the title in the database and someone later changes the product title in the WC store, will the changes sync quickly? If not, we might run into issues with the DB approach. What do you think?

mukeshpanchal27 replied:
Update in https://github.com/woocommerce/google-listings-and-ads/pull/2935/commits/4ada6dff61130bc4904e1d076bcda15f08a620f6

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -101,15 +310,145 @@ public function get_summary(): array {
+			'orderby' => 'mc_insights_effectiveness',
+		];
+
+		$args = array_merge( $defaults, $args );
```
You can use `wp_parse_args()` here.

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -101,15 +310,145 @@ public function get_summary(): array {
+		}
+
+		// Set order and orderby.
+		$order   = strtoupper( $args['order'] ) === 'ASC' ? 'ASC' : 'DESC';
+		$orderby = ! empty( $args['orderby'] ) ? $args['orderby'] : 'id';
```
Since we're defining the defaults above, we should not need fallbacks here.

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -21,6 +22,196 @@
+	 * @param array $args Query arguments.
+	 * @return array Mapped response data.
+	 */
+	private function get_price_benchmarks_base_data( $args ) {
```
Now that we're merging data from all 3 Content API queries, we no longer need to separate "base data" from the merchant performance data. Let's fetch all 3 sources and pass them all to the `map_price_benchmarks_response()` method to format the data for inserting into the store DB.

mukeshpanchal27 replied:
Addressed in https://github.com/woocommerce/google-listings-and-ads/pull/2935/commits/b52ed875410801918bee3071cd67b49cbb33d199

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -21,6 +22,196 @@
+	 * @param int|string $id The product ID to fetch performance data for.
+	 * @return array Associative array with 'clicks' and 'conversions' keys, or empty array on failure.
+	 */
+	public function get_performance_data( $id ): array {
```
This method shouldn't be needed if we can call `$merchant->get_merchant_performance_data()` directly, similar to how we are calling

```
$benchmark_data = $merchant->get_price_comparisons_data( $args );
$price_insights_data = $merchant->get_price_insights_data( $args );
```

joemcgill commented on src/MerchantCenter/PriceBenchmarks.php:
```diff
@@ -101,15 +310,145 @@ public function get_summary(): array {
 			$job->schedule();
```
Now that we're relying on the local DB being up to date for both the summary data and for the price benchmarks list, we should look at moving the responsibility for scheduling this job elsewhere and make this method only responsible for getting and formatting the summary data.

mukeshpanchal27 replied:
Instead, can we schedule the job during plugin load? Since it's a recurring event, it will automatically reschedule itself once the first execution is completed.

joemcgill replied:
I'm having a bit of trouble fully understanding how all of the pieces of the Jobs system is meant to work, but found that the `JobInitializer` class [has some special handling ](https://github.com/woocommerce/google-listings-and-ads/blob/c5c67ec46c7ad7218ecc554edae7e8e952375712/src/Jobs/JobInitializer.php#L43-L72)for certain types of jobs to register a hook for scheduling and another that automatically calls that hook once a day.

I've made that change in 6888aee and it seems to be working.

mukeshpanchal27 replied:
Seems it set schedule job 3AM everyday.

https://github.com/user-attachments/assets/36ddc20d-7d50-41ed-9898-b7e5312b9bd7

joemcgill replied:
Correct. It seems like any jobs implementing the `RecurringJobInterface` will automatically schedule a 'start' event to run every day at 3am, but that event is only used to ensure the job is scheduled.

<img width="1509" alt="image" src="https://github.com/user-attachments/assets/915304b1-ee67-4523-85e0-f758941e22a3" />

You can see here that there are two scheduled jobs:

* `gla/jobs/update_merchant_price_benchmarks/start` calls the start hook which causes the `UpdateMerchantPriceBenchmarks::schedule()` method to be called.
* `gla/jobs/update_merchant_price_benchmarks/process_item` runs every 12 hours and calls the `UpdateMerchantPriceBenchmarks::process_items()` method.

joemcgill commented on src/API/Site/Controllers/MerchantCenter/PriceBenchmarksController.php:
```diff
@@ -94,17 +94,18 @@ protected function prepare_query_arguments( Request $request ): array {
+				/** @var PriceBenchmarks $price_benchmarks */
+				$price_benchmarks = $this->container->get( PriceBenchmarks::class );
+
+				$response_data = $price_benchmarks->get_price_benchmarks_data();
```
We'll still need to accept and prepare query args here and use those as the args for the get_price_benchmarks_data method.
```suggestion
				$response_data = $price_benchmarks->get_price_benchmarks_data( $this->prepare_query_arguments( $request ) );
```

joemcgill reviewed (commented):
@mukeshpanchal27 good progress so far. I've left some initial feedback.

When you do these updates, can you add some testing instructions to help confirm that both the data is being properly added to the database from the Google APIs and that all of the required query params we need to support are being accounted for when we retrieve the data from the local DB?

mukeshpanchal27 committed:
Merge branch 'feature/2824-price-benchmarks' into update/2911-codebase

mukeshpanchal27 committed:
Update docblock

mukeshpanchal27 committed:
Use product helper function instead of polyfill

mukeshpanchal27 committed:
Address review feedbacks

mukeshpanchal27 committed:
Use wp_parse_args instead of array_merge

mukeshpanchal27 committed:
Remove unwanted fallbacks

mukeshpanchal27 committed:
Remove base data function

mukeshpanchal27 committed:
Add args support

mukeshpanchal27 committed:
Add product search

mukeshpanchal27 committed:
Add special id  support in get_wc_product_id()

mukeshpanchal27 committed:
Add special id  support in get_wc_product_id()

mukeshpanchal27 committed:
Update unit tests

mukeshpanchal27 commented:
@joemcgill I'm having trouble figuring out how to mock the `get_price_benchmarks_response` method in the tests. If you have a moment, could you please help by adding the relevant test coverage?

joemcgill committed:
Make 'UpdateMerchantPriceBenchmarks' job auto-start

mukeshpanchal27 commented:
@joemcgill I have updated the Testing instruction for the PR.

joemcgill committed:
Handle standard REST API args in PriceBenchmarks::get_price_benchmarks_data

joemcgill committed:
Normalize case handling for asc/desc order attributes

joemcgill committed:
Remove unused code

joemcgill committed:
Add PHPUnit tests for PriceBenchmarks

jjgrainger commented on src/API/Site/Controllers/MerchantCenter/PriceBenchmarksController.php:
```diff
@@ -96,15 +96,16 @@ protected function prepare_query_arguments( Request $request ): array {
+
+				$response_data = $price_benchmarks->get_price_benchmarks_data( $this->prepare_query_arguments( $request ) );
 
 				if ( empty( $response_data ) ) {
 					return new Response( [] );
 				}
```
@mukeshpanchal27 @joemcgill very minor thing but is this check necessary? If the `$response_data` is empty is it not the same as the following line `return new Response( $response_data )`?

joemcgill replied:
Assuming the empty response is an empty array and not something like `null` or `false`, then yes. I think this just ensures we're sending the correct shape. The `PriceBenchmarks::get_price_benchmarks_data()` method has strict typing that it returns an array, so it's probably fine to remove this.

joemcgill replied:
Addressed in [1583097](https://github.com/woocommerce/google-listings-and-ads/pull/2935/commits/15830976ecdf0e9c418dd483310542d7d9fd3069)

jjgrainger commented on src/DB/Table/MerchantPriceBenchmarksTable.php:
```diff
@@ -56,14 +67,25 @@ public static function get_raw_name(): string {
-			'benchmark_price_micros'        => true,
-			'benchmark_price_currency_code' => true,
-			'price_compared_with_benchmark' => true,
+			'product_id'                                   => true,
+			'mc_product_id'                                => true,
+			'mc_product_offer_id'                          => true,
+			'mc_product_price_micros'                      => true,
+			'mc_product_currency_code'                     => true,
+			'mc_price_country_code'                        => true,
+			'mc_price_benchmark_price_micros'              => true,
+			'mc_price_benchmark_price_currency_code'       => true,
+			'mc_insights_suggested_price_micros'           => true,
+			'mc_insights_suggested_price_currency_code'    => true,
+			'mc_insights_predicted_impressions_change_fraction' => true,
+			'mc_insights_predicted_clicks_change_fraction' => true,
+			'mc_insights_predicted_conversions_change_fraction' => true,
+			'mc_insights_effectiveness'                    => true,
+			'mc_metrics_clicks'                            => true,
+			'mc_metrics_impressions'                       => true,
+			'mc_metrics_ctr'                               => true,
+			'mc_metrics_conversions'                       => true,
+			'price_compared_with_benchmark'                => true,
```
Looks like this needs formatting
```suggestion
			'product_id'                                        => true,
			'mc_product_id'                                     => true,
			'mc_product_offer_id'                               => true,
			'mc_product_price_micros'                           => true,
			'mc_product_currency_code'                          => true,
			'mc_price_country_code'                             => true,
			'mc_price_benchmark_price_micros'                   => true,
			'mc_price_benchmark_price_currency_code'            => true,
			'mc_insights_suggested_price_micros'                => true,
			'mc_insights_suggested_price_currency_code'         => true,
			'mc_insights_predicted_impressions_change_fraction' => true,
			'mc_insights_predicted_clicks_change_fraction'      => true,
			'mc_insights_predicted_conversions_change_fraction' => true,
			'mc_insights_effectiveness'                         => true,
			'mc_metrics_clicks'                                 => true,
			'mc_metrics_impressions'                            => true,
			'mc_metrics_ctr'                                    => true,
			'mc_metrics_conversions'                            => true,
			'price_compared_with_benchmark'                     => true,
```

joemcgill replied:
Agree that your suggestion looks better. I've noticed that once array keys get to a certain length, the alignment WPCS rules no longer apply. If you do apply these changes locally, they get auto-formatted back to the way they were previously. Kind of odd?

joemcgill replied:
Looks like this is a setting you can modify in the ruleset. I've increased the max columns in [58f4605](https://github.com/woocommerce/google-listings-and-ads/pull/2935/commits/58f4605712da5e6958cbc68f902dd1823a315803)

jjgrainger reviewed (commented):
@mukeshpanchal27 @joemcgill overall looks good, including the new tests.

I've left some minor comments to review. Thanks

joemcgill committed:
Remove redundant empty array check

joemcgill committed:
Increase WPCS max alignment columns

joemcgill committed:
Small cleanup

joemcgill reviewed (dismissed):
Pushed a few more cleanup changes but this looks good to merge

joemcgill committed:
Update orderby support for Price Benchmarks

This updates the PriceBenchmarks controller to specify a supported map of orderby values and maps them to the appropriate fields in the database query. The default value is also set as a constant in the class which is then referenced in the PriceBenchmarksController class that defines the API endpoint.

Unit tests are added to test all the supported orderby values and ensure that the API behaves as expected.

joemcgill committed:
Update API schema and improving typing

joemcgill committed:
Update API response and schema shape

joemcgill commented:
@jjgrainger I've made [a couple of changes](https://github.com/woocommerce/google-listings-and-ads/pull/2935/files/bff5e19d663ef289597f61c1d5d5ea3a324b55d6..5c3425e87ce361758e8d4d3a4a074895c82dcf33) since your last review because I realized that the way some of this was implemented was based on an incorrect data shape. I've gone ahead and made the required changes and updated some unit tests, but would appreciate another set of eyes.

jjgrainger reviewed (commented):
Thanks @joemcgill I've reviewed and the changes look good

joemcgill reviewed (approved):
Thanks @jjgrainger. I'm going to go ahead and merge this to the feature branch so we can test the FE changes integrated with the backend.

ankitguptaindia commented:
**QA Status-** Approved ✅ 
------------

**Testing Environment -** 

<Details>

* WordPress: 6.8.1
* Theme:  
  - Twenty Twenty-Five Version: 1.2
* WooCommerce - Version 9.9.0-rc.1
* PHP: 8.3.0
* Web Server: Nginx
* Browser: Chrome - Version 137
* OS: macOS Sonoma 14.6.1 

</Details>




**Test Results -** 

Tested the product listing flow for effectiveness, sorting by using the steps as described. Products appear correctly sorted by effectiveness in **descending order** (i.e., highest to lowest).
Also tested the **change in sorting order** — both **descending and ascending** — and verified that sorting works correctly even when pagination is active on the second page.



**Functional Demo /  Screencast -**. 


https://github.com/user-attachments/assets/830447f9-3bc0-4f8c-9d28-eb6569fe3432
